service/history/ndc/workflow_state_replicator.go

```diff
@@ -187,7 +198,551 @@ func (r *WorkflowStateReplicatorImpl) SyncWorkflowState(
+	namespaceID namespace.ID,
+	workflowID string,
+	originalMutableState workflow.MutableState,
+	request *shard.ReplicateMutableStateRequest,
+) (Workflow, error) {
+	newMutableState, err := r.getNewRunMutableState(ctx, namespaceID, workflowID, request.NewRunInfo.GetRunId(), originalMutableState, request.NewRunInfo.EventBatch)
+	if err != nil {
+		return nil, err
+	}
+	newExecutionInfo := newMutableState.GetExecutionInfo()
+	newExecutionState := newMutableState.GetExecutionState()
+	newContext := workflow.NewContext(
+		r.shardContext.GetConfig(),
+		definition.NewWorkflowKey(
+			newExecutionInfo.NamespaceId,
+			newExecutionInfo.WorkflowId,
+			newExecutionState.RunId,
+		),
+		r.logger,
+		r.shardContext.GetThrottledLogger(),
+		r.shardContext.GetMetricsHandler(),
+	)
+
+	return NewWorkflow(
+		r.clusterMetadata,
+		newContext,
+		newMutableState,
+		wcache.NoopReleaseFn,
+	), nil
+}
+
+func (r *WorkflowStateReplicatorImpl) getNewRunMutableState(
+	ctx context.Context,
+	namespaceID namespace.ID,
+	workflowID string,
+	newRunID string,
+	originalMutableState workflow.MutableState,
+	newRunEventsBlob *commonpb.DataBlob,
+) (workflow.MutableState, error) {
+	newRunHistory, err := r.historySerializer.DeserializeEvents(newRunEventsBlob)
+	if err != nil {
+		return nil, err
+	}
+	sameWorkflowChain := false
+	newRunFirstEvent := newRunHistory[0]
+	if newRunFirstEvent.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED {
+		newRunFirstRunID := newRunFirstEvent.GetWorkflowExecutionStartedEventAttributes().FirstExecutionRunId
+		sameWorkflowChain = newRunFirstRunID == originalMutableState.GetExecutionInfo().FirstExecutionRunId
+	}
+
+	var newRunMutableState workflow.MutableState
+	if sameWorkflowChain {
+		newRunMutableState, err = workflow.NewMutableStateInChain(
+			r.shardContext,
+			r.shardContext.GetEventsCache(),
+			r.logger,
+			originalMutableState.GetNamespaceEntry(),
+			workflowID,
+			newRunID,
+			timestamp.TimeValue(newRunHistory[0].GetEventTime()),
+			originalMutableState,
+		)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		newRunMutableState = workflow.NewMutableState(
+			r.shardContext,
+			r.shardContext.GetEventsCache(),
+			r.logger,
+			originalMutableState.GetNamespaceEntry(),
+			workflowID,
+			newRunID,
+			timestamp.TimeValue(newRunHistory[0].GetEventTime()),
+		)
+	}
+
+	newRunStateBuilder := workflow.NewMutableStateRebuilder(r.shardContext, r.logger, newRunMutableState)
+
+	_, err = newRunStateBuilder.ApplyEvents(
+		ctx,
+		namespaceID,
+		uuid.New(),
+		&commonpb.WorkflowExecution{
+			WorkflowId: workflowID,
+			RunId:      newRunID,
+		},
+		[][]*historypb.HistoryEvent{newRunHistory},
+		nil,
+		"",
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return newRunMutableState, nil
+}
+
+// this function does not handle reset case, so it should only be used for the case where workflow run is found at local cluster
+func (r *WorkflowStateReplicatorImpl) bringLocalEventsUpToSourceCurrentBranch(
```

Comment:
for future: we may need to some checkpoint mechanism for backfilling to handle large histories. One idea can be: create a temp branch in version histories and use that to record how many events have been backfilled.